ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+# Ruby SDK (lightweight overview)
+
+This is a lightweight proposal for the Ruby SDK. The intention for this proposal is to give everyone
+an end-to-end feel for the Ruby SDK. Many aspects of this proposal will probably change in the more
+detailed versions later.
+
+
+## Types
+
+After many experiments we've decided to not proceed with Ruby typing at this point. This decision
+will be revisited when Ruby's typing system matures to support non-basic use-cases.
+
+
+## Defining an activity
+
+Activities are defined as classes ([a very common practice](https://guides.rubyonrails.org/active_job_basics.html)
+given that the language is object-oriented). Activities are expected to implement an `#execute`
+method that will be called to execute an activity.
+
+`sum_activity.rb`
+
+```ruby
+# The activity class name will be used as activity name unless overridden
+class SumActivity < Temporal::Activity
+  # Optionally override the name
+  activity_name 'sum-activity'
+
+  # return value of the #execute method will be used as an activity result
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+#### Alternative approaches
+
+- Using plain methods as activities. This is not feasible in Ruby since non-namespaced methods
+  are global without a clear way to reference them (would have to be `method(:my_activity)`)
+
+- It is possible to allow multiple activities per class, however we've decided against this because
+  there's an extra complexity to the invocation (extra argument representing the method) and it
+  complicates the use of class-level "decorators" (e.g. for specifying activity name or other
+  options)
+
+
+## Defining a workflow
+
+Similar to activities, workflows are also represented as classes.
+
+`sum_workflow.rb`
+
+```ruby
+require 'sum_activity'
+
+# Similar to activities the class name will be used as a workflow name unless overridden
+class SumWorkflow < Temporal::Workflow
+  # Optionally override the name
+  workflow_name 'sum-workflow'
+
+  def execute(a, b)
+    # - Positional arguments passed to activity (inline with other SDKs)
+    # - Keyword arguments are reserved for options
+    # - This call is blocking
+    workflow.execute_activity(SumActivity, a, b, start_to_close_timeout: 10)
```

Comment:
Also, because you're going with one-activity-per-class (yay!), it's great if the activity author specifies the timeouts and retry policies rather than the caller, because the activity author is the expert (often they are two different people). In that world, the caller only overrides them in in rare cases, which may make it easier for you to stomach an explicit options class rather than the pattern of keyword-args-as-options.

FYI, if you're not aware, Temporal is working on adding ExecuteAtMostOnce vs ExecuteAtLeastOnce annotations to the SDKs.
(see https://temporalio.notion.site/Replay-roadmap-highlights-print-version-632692bcd46549719f91bb9b7ac859bd "shorthand retry policies")

This is another aspect that pushes us toward wanting more stuff declared on the activity.

Reply:
Due to how Temporal API works and that activities can be called across languages, retry policies and timeouts must be caller-time values. You can't know a Ruby activity's retry policy when called from Go.

However, we encourage people to use normal code sharing practices so _they_ can centralize their retry policies. So if you're a Ruby-only shop you can put them on this class and reference them, but if you're not maybe you'll have them in a proto contract or just known on the caller side.

Most users agree that definition-time activity options make sense, but the core primitives of Temporal don't support that yet.

Reply:
We are trying to keep from having two different ways to provide the options to an activity at this low level. But this is really easy with just a constant on the activity they can reference. We are also trying to keep Ruby from being special here compared to all other SDKs.

Reply:
@drewhoskins-stripe I agree with you on this and in fact this is how the current SDK behaves allowing you to specify it on the Activity class and then optionally override when calling an Activity. This is a step away from the other SDKs, but I think that the key is extensibility of the SDK allowing you to easily add support for this behaviour (for example via an interceptor) and maybe publish it as a gem.

Using a single class to combine timeouts and retry policy options does make it easier (although admittedly forces more typing):

```ruby
workflow.execute_activity(SumActivity, a, b, timeouts: Temporal::Timeouts.new(start_to_close: 10), retry_policy: Temoral::RetryPolicy.new(...))
```

@cretz @bergundy @Sushisource what do you think? 

@cretz I think the cross-language argument doesn't really apply here because these calls similarly can't know remote argument/result serialization. But that doesn't mean we don't want to support it for same-language use-cases and via a stub for remote.

Reply:
> what do you think?

Does Ruby offer a splat/spread operator? I don't think there is much value in a "Timeouts" grouping. But there may be value in an "ActivityOptions" grouping that you can splat, clone/override, etc.

> I think the cross-language argument doesn't really apply here because these calls similarly can't know remote argument/result serialization.

To clarify, we are talking about definition-site activity options. Invocation from non-Ruby won't have access to those options. You might as well just make them a constant on the class and reference them at call time instead of having it built in. I don't think there is enough benefit in having an official Temporal way for declaration-site options that only half-work and have no server support and make the Ruby SDK special.

I think tackling declaration-site options is a reasonable thing to do holistically, not shoehorned into a not-yet-built SDK. We know we need call-site options anyways, lets get that built first for parity purposes.

Reply:
I don't think we should group timeouts. They are often unrelated to each other. It's not like a "timeout policy".